### ISSUE TITLE
Optional properties and object spread don't work properly with functional stateless components

### ISSUE THREAD

**Issue by nthtran** [facebook/flow#1878]
An example:

``` js
type UserProps = {
  name?: string,
};

const User = ({ name }: UserProps) =>
  <div>{name}</div>;

type AppProps = {};

export const App = (props: AppProps) =>
  <User {...props} name="hello" />;
```

This results in `undefined This type is incompatible with string` on line 2.

If the `{...props}` is removed then there's no error. If `User` is a `React.Component`instead of a functional component then there's no error as well:

``` js
class User extends React.Component {
  props: UserProps;

  render() {
    const { name } = this.props;
    return <div>{name}</div>;
  }
}
```

There's also no error if `name` is not optional on `UserProps`:

``` js
type UserProps = {
  name: string,
};
```

**Comment by jedwards1211** [comment]
I just upgraded from Flow 0.23.0 to 0.30.0 and started seeing this, I think it's a regression.

**Comment by whw** [comment]
I'm seeing this also on 0.32.0

**Comment by chris-lock** [comment]
+1

**Comment by nthtran** [comment]
It seems this has been fixed in `flow@0.34.0`. Closing.
